integration_tests/main.rs

```diff
@@ -0,0 +1,276 @@
+            operations: vec![(
+                Operation::Establish,
+                Outcome {
+                    customer_status: CustomerStatus::Ready,
+                    merchant_status: MerchantStatus::Active,
+                    error: None,
+                },
+            )],
+        },
+        Test {
+            name: "Channels cannot share names".to_string(),
+            operations: vec![
+                (
+                    Operation::Establish,
+                    Outcome {
+                        customer_status: CustomerStatus::Ready,
+                        merchant_status: MerchantStatus::Active,
+                        error: None,
+                    },
+                ),
+                (
+                    Operation::Establish,
+                    Outcome {
+                        customer_status: CustomerStatus::Ready,
+                        merchant_status: MerchantStatus::Active,
+                        error: Some(Party::ActiveOperation("establish")),
+                    },
+                ),
+            ],
+        },
+    ]
+}
+
+#[derive(Debug, Error)]
+enum TestError {
+    #[error("Operation {0:?} not yet implemented")]
+    NotImplemented(Operation),
+
+    #[error(
+        "The error behavior did not satisfy expected behavior {op:?}. Got
+    CUSTOMER WATCHER OUTPUT:
+    {customer_errors}
+    MERCHANT SERVER OUTPUT:
+    {merchant_errors}
+    OPERATION OUTPUT:
+    {op_error:?}"
+    )]
+    InvalidErrorBehavior {
+        op: Operation,
+        customer_errors: String,
+        merchant_errors: String,
+        op_error: Result<(), anyhow::Error>,
+    },
+}
+
+impl Test {
+    async fn execute(
+        &self,
+        rng: &StdRng,
+        customer_config: &customer::Config,
+        merchant_config: &merchant::Config,
+    ) -> Result<(), anyhow::Error> {
+        for (op, expected_outcome) in &self.operations {
+            // Clone inputs. A future refactor should look into having the `Command` trait take
+            // these by reference instead.
+
+            let outcome = match op {
+                Operation::Establish => {
+                    let est = customer_cli!(
+                        Establish,
+                        vec![
+                            "establish",
+                            "zkchannel://localhost",
+                            "--label",
+                            &self.name,
+                            "--deposit",
+                            "5 XTZ"
+                        ]
+                    );
+                    est.run(rng.clone(), customer_config.clone())
+                }
+                Operation::NoOp => Box::pin(async { Ok(()) }),
+                err_op => return Err(TestError::NotImplemented(*err_op).into()),
+            }
+            .await;
+
+            // Sleep until the servers have finished their thing, approximately
+            tokio::time::sleep(op.wait_time()).await;
```

Comment:
Can we do pulling of logs (with timeout) instead?

Reply:
I can try this. My concerns are:
- There are a lot of various correct "done" scenarios. I'm hesitant about hard-coding a "done" message for every `Operation` into zeekoe, especially when there are multiple valid paths. I would be concerned about missing one.
- I'm not sure what would happen if we find an actual failure case (e.g., where the server does not get to the "done" message because something errored). We would still need a timeout larger than the max expected time for any of these things to run.

Reply:
I just added some infra in 2320d5c to wait for specific logs - in that case, to see if the merchant server and customer watcher are up. I can reuse a lot of that for this situation too.

Reply:
I'm getting stuck on order of operations and correctness conditions here. For example, I would need to check the merchant server log for a "successfully established" log. But if the channel was supposed to fail to establish, then I need to skip that. But I don't want to check whether it was supposed to fail until after I know the server is done. 

I think I am going to stick with the timeout approach for now and we can discuss how this could work in a call at some point.